### PR TITLE
Visitor can sign up

### PR DESCRIPTION
**Pull request by perhegg**
## Feature: User can register an account
https://www.pivotaltracker.com/story/show/163901539

### User Story:
* As a user,
    In order to have be able to finalize an order
    I would like to be able to register an account

### Description of PR:
* Creation of user model
* Add factory bot for creating user
* Tests created for signing up

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,2 +1,13 @@
 class ApplicationController < ActionController::Base
+protect_from_forgery with: :exception
```
Comment (vick3d) [inline app/controllers/application_controller.rb]: indentation

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,2 +1,13 @@
...
+
+before_action :configure_permitted_parameters, if: :devise_controller?
```
Comment (vick3d) [inline app/controllers/application_controller.rb]: indentation

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,2 +1,13 @@
...
+
+protected
+
+def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:email])
```
Comment (vick3d) [inline app/controllers/application_controller.rb]: I'm not sure if we need to add email as a key, isn't email already included as default from devise?

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -1,2 +1,13 @@
...
+    devise_parameter_sanitizer.permit(:account_update, keys: [:email])
```
Comment (vick3d) [inline app/controllers/application_controller.rb]: same as above

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+class User < ApplicationRecord
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+
+         validates :email, length: { maximum: 20 }
```
Comment (vick3d) [inline app/models/user.rb]: indentation?

**Review by vick3d — CHANGES_REQUESTED** [review]
We also need to look at the merge conflicts

File: features/step_definitions/basic_steps.rb [hunk 6]
```diff
@@ -1,14 +1,21 @@
+When("I click on {string}") do |button_name|
+    click_on button_name
+end
+
+When("I fill in {string} with {string}") do |field, input|
+    fill_in field, with: input
+end
   Given("the following products exists") do |table|
```
Comment (vick3d) [inline features/step_definitions/basic_steps.rb]: identation

**Review by vick3d — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (LauraRe) [reply app/controllers/application_controller.rb]: fixed!

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (LauraRe) [reply app/controllers/application_controller.rb]: fixed!

File: app/controllers/application_controller.rb (hunk 3 above)
Comment (LauraRe) [reply app/controllers/application_controller.rb]: Took it away.

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (LauraRe) [reply app/controllers/application_controller.rb]: done

File: app/models/user.rb (hunk 5 above)
Comment (LauraRe) [reply app/models/user.rb]: updated, indentation fixed

**Review by vick3d — APPROVED** [review]
LGTM! :)

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -1,2 +1,12 @@
 class ApplicationController < ActionController::Base
-end
+    protect_from_forgery with: :exception
+
+    before_action :configure_permitted_parameters, if: :devise_controller?
+
+    protected
+
+    def configure_permitted_parameters
+        devise_parameter_sanitizer.permit(:sign_up)
```
Comment (oliverochman) [inline app/controllers/application_controller.rb]: why have you added this?

File: app/views/devise/confirmations/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,16 @@
+<h2>Resend confirmation instructions</h2>
```
Comment (oliverochman) [inline app/views/devise/confirmations/new.html.erb]: do you need to have all of these devise views in the application?
You only need them if you are going to modify them

File: app/views/products/index.html.haml [hunk 9]
```diff
@@ -1,5 +1,9 @@
 %h1 Peanuts
 
+<header>
```
Comment (oliverochman) [inline app/views/products/index.html.haml]: please dont mix haml with regular html code, keep it consistent

File: config/initializers/devise.rb [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+  # random tokens. Changing this key will render invalid all existing
+  # confirmation, reset password and unlock tokens in the database.
+  # Devise will use the `secret_key_base` as its `secret_key`
+  # by default. You can change it below and use your own secret key.
+  # config.secret_key = 'c22ccbfd3e2107fa26494f66e35d57e86b1fb3d57c613fbe9692c3605e44a676e3b66a1669ab39cfc2bb66a5ba531d34b0221f2c7fdef56e5cb3e4438ab8dd23'
+
+  # ==> Controller configuration
+  # Configure the parent class to the devise controllers.
+  # config.parent_controller = 'DeviseController'
+
```
Comment (oliverochman) [inline config/initializers/devise.rb]: clean this file up

File: features/signup.feature [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    And I fill in "Email" with "laura@gmail.com"
+    And I fill in "Password" with "0443431157"
+    And I fill in "Password confirmation" with "0443431157"
+    And I click on "Sign up"
+    Then I should see "Welcome! You have signed up successfully." 
+
+Scenario: User visits page and enters incorrect Password Confirmation [SadPath]
+    When I click on "Sign up" 
+    And I fill in "Email" with "laura@gmail.com"
+    And I fill in "Password" with "0443431157"
```
Comment (oliverochman) [inline features/signup.feature]: can you make it more obvious why the sign up in the sad path does not work?

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 7 above)
Comment (perhegg) [reply app/controllers/application_controller.rb]: we are going to need it later when we implement cartify with more parameters ("name" for example).

Comment (oliverochman) [reply app/controllers/application_controller.rb]: Premature optimization. Need to stay in scope. Please remove it if it has nothing to do with the feature

File: app/views/devise/registrations/edit.html.erb [hunk 12]
```diff
@@ -0,0 +1,43 @@
+<h2>Edit <%= resource_name.to_s.humanize %></h2>
```
Comment (oliverochman) [inline app/views/devise/registrations/edit.html.erb]: If you dont modify these devise view files you need to remove them

File: spec/models/user_spec.rb [hunk 13]
```diff
@@ -0,0 +1,20 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  describe 'DB table' do
+    it { is_expected.to have_db_column :id }
```
Comment (oliverochman) [inline spec/models/user_spec.rb]: there is no need to test for `:id`, every model you create in rails have ids

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 7 above)
Comment (perhegg) [reply app/controllers/application_controller.rb]: removed.

File: app/views/devise/confirmations/new.html.erb (hunk 8 above)
Comment (perhegg) [reply app/views/devise/confirmations/new.html.erb]: removed

File: app/views/devise/registrations/edit.html.erb (hunk 12 above)
Comment (perhegg) [reply app/views/devise/registrations/edit.html.erb]: removed

File: spec/models/user_spec.rb (hunk 13 above)
Comment (perhegg) [reply spec/models/user_spec.rb]: removed testing for id for user and product

**Review by oliverochman — APPROVED** [review]
Good job guys

**Pull request merged**